### ISSUE TITLE
Affine Constraints and Loop Optimizations

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#380]
I have a philosophical question about the nature of constraints that I came up on while adding support for affine analysis to Regent.

In particular, what I added was the ability for Regent to determine that `p[i]` and `p[i+1]` are disjoint in the code below (assuming `p` is disjoint):

```
for i in whatever do
   some_task(p[i], p[i+1])
end
```

The straightforward way to do this is to just add a constraint that `p[i] * p[i+1]`. This is powerful because it means that anywhere Regent wants to know a disjointness constraint, this can be used. However, it also introduces a pretty severe soundness bug into any loop optimizations that rely on constraints. In particular, in the loop above, `p[i]` and `p[i+1]` are disjoint, so every individual task call is valid, but the loop as a whole cannot be optimized into an index launch because different iterations are still disjoint. This caused me to question if my definition of "constraint" is correct.

The implicit definition used to justify the constraint above might be stated:

 1. A constraint is a logical proposition that holds at every program point where the referenced variables are in scope.

This makes sense for the constraint `p[i] * p[i+1]` because it holds for any given value of `i` (when `i` is in scope). On the other hand, you could imagine a definition like:

 2. A constraint is a proposition where the free variables can be independently instantiated with any values those variables can hold any any point in the program execution where those variables are in scope.

So then `p[i] * p[i+1]` would not hold because it's possible to come up with two instantiations of `i` called `i1` and `i2` such that `p[i1] * p[i2+1]` is not true.

The second definition is more like what we need for a loop optimization, but I think it's not expressive enough to capture cases we need for DOM in Soleil. Specifically, I came up with definition (1) specifically to enable the following use case:

```
for d in diagonals do
    for t in tiles do
        some_task(p[d][t], p[d+1][t])
    end
end
```

In order to be able to index launch the inner loop, we need to know that `p[d]` and `p[d+1]` are disjoint. In this case, relying on such a constraint is valid, because `d` does not vary across the inner loop.

So I'm starting to wonder if the correct answer is to go with definition (1), since it's so useful, and to somehow tag constraints with the variables they're associated with so that loop optimizations like index launches can ignore certain sets of constraints based on what variables are known to vary across iterations the loop. But this also seems a bit hacky, and I'd be happy to find a cleaner solution.

**Comment by lightsighter** [comment]
First off, I thought Regent currently isn't doing any static checking of non-interference for regions in a task launch, at least that was my assumption when I created issue #195. If we're not doing that at all right now, I feel like we can continue just ignoring interference of region requirements and let the runtime deal with that, and only worry about checking for dependences between loop iterations so we can determine when we can convert the loop into an index space launch.

Second, why did you write the loop for DOM like that? The explicit "diagonal" is really a performance crutch since we can't do constrained index space launches yet. If we remove that restriction though, then the natural way to write I would think would be this:
for x in xbounds do 
    for y in ybounds do 
         some_task(p[x-1][y], p[x][y-1], p[x][y]) -- ignore boundaries with this indexing
    end
end
The compiler should be able to use the syntax as input to @TWarszawski constraint solver to figure out the constraints on what can be run in parallel and whether we can launch this as an index space launch or not. That's at least how it should work eventually.

As for how it works now, I think we should do whatever we need to in order to get that inner loop to turn into an index space launch. I wouldn't bother with the interference checking for individual region requirements in a task since we already aren't handling that in Regent anyway (unless I'm missing something and issue #195 should be closed).

**Comment by karthikmurthy** [comment]
I actually think that the scope controls the free-variable/parameter definition. At the scope of your t-loop, i.e., when you are reasoning about the index-launch of t-loop, the assumption that d is a free-variable is not true, it is in scope but it is now a parameter to the solver. In this case, the reasoning will be solely based on  values that t can take and hence, will the iterations independent, allowing for index launch. 

Or, if you specify as two free variables like what @mbauer did, then @TWarszawski <https://github.com/TWarszawski> solver should be able to extract diagonal iterations that are independent, based on what I know of his solver.

**Comment by elliottslaughter** [comment]
#195 is for single task launches. We have always had non-interference checks for index launches. This is a feature; for individual task launches we want to be permissive, but for index launches we only want to do the optimization if we can guarantee that it'll work.

I think we need ordering constraints on the loops to make the loop @lightsighter suggested work. The way I'd write it is:

```
for xy in xybounds do
    some_task(p[xy - { 1, 0 }], p[xy - { 0, 1}], p[xy])
end
```

I'd prefer that users use explicitly multi-dimensional loops, because I explicitly don't intend to consider loop nests for index launches.

I see using @TWarszawski's constraint solver as an implementation detail. My one hard constraint is that we absolutely must not depend on Legion runtime code at compile time. When we're cross-compiling that's just not going to work. In fact I'd rather avoid depending on C++ code altogether, because it's just a hairy mess even without pulling in the entire Legion code base. So my intention was to build whatever I'm going to build in the compiler, and have it be separate from what we do for code generation to target the new projection functor API.

I think @karthikmurthy is probably on the right track. If we consider only variables that are valid outside of the loop that's currently being analyzed, we'll naturally avoid any unsound constraints. This still requires tracking the variables that are associated with constraints, which means revising the compiler internals.

**Comment by elliottslaughter** [comment]
Sorry, I got that loop wrong. Should have been:


```
for xy in xybounds order by xy.x + xy.y do
    some_task(p[xy - { 1, 0 }], p[xy - { 0, 1}], p[xy])
end
```

**Comment by magnatelee** [comment]
I agree with @karthikmurthy that variable `d` should be considered as a fixed variable in the innermost loop and the constraint should not escape that scope. I guess the issue is that the Regent compiler has been dealing with constraints in flow-insensitive, scope-insensitive manner, and now it needs to be made sensitive at least to scopes, which I think isn't too hard.

BTW, I want to point out that @debnil has been implementing aliased region requirement checks for Regent, which will partially address #195 and he will need to extend his checks for non-identity projection functors as well.

**Comment by alexaiken** [comment]
I've only read Elliott's original post, so maybe this has already been addressed in subsequent comments --- apologies if these comments are redundant.

Cut and paste of the nested loop example:

for d in diagonals do
    for t in tiles do
        some_task(p[d][t], p[d+1][t])
    end
end

I think there are two separate issues.  The first is that quantifiers can be used to make the scopes and the kind of question we are asking.  So, for example, if we want to check whether the inner loop can be parallelized, the right form of the constraint is something like

\forall t. ... expression in t and d ...

This explicitly leaves d free and makes it clear it is an unknown constant for the purposes of this query.  In general, for a query over n inner loops, the inner n loop variables are quantified and the outer n are left free.  This is isn't quite enough, because ...

The second part is how we express whether two loop iterations might conflict.  I think what we want to express is:  No two different loop iterations conflict on a data access.  So the form of the constraint above for the inner loop would be something like

\forall t.  \forall t'.   t = t' or ((p[d][t] != p[d][t'+1]) and (p[d][t] != p[d][t']) and (p[d][t+1] != p[d][t']) and (p[d][t+1] != p[d][t'+1})

i.e., we explicitly quantify over two loop iterations t and t', the statement is vacuous if t = t', and if they are different we check whatever the condition is for that loop.

I think this is the right semantic framework, but you probably don't want something this complicated to be what programmers use.  A well-defined translation from whatever surface syntax is used to a richer logic internally that can make all the right distinctions seems like a good way to go to me.

**Comment by elliottslaughter** [comment]
The soundness hole should be fixed in 6d0f056023dd319f2fc2ca33ba606aecc7be9886. Thanks everyone for your input.
